### PR TITLE
Signup: Force a re-rendering of translations when the locale changes in props

### PR DESCRIPTION
**Pull request by kwight**
Using the "Also available in..." prompt in signup to switch languages doesn't work properly. For example, from FR to EN works the first time, but not the second (going back to FR).

screen shot 2016-03-29 at 5 00 05 pm

screen shot 2016-03-29 at 5 00 29 pm

screen shot 2016-03-29 at 5 00 46 pm

### CODE REVIEW

**Comment by kwight** [comment]
@sirbrillig can you see if there's anything we discussed that I've missed perhaps?

**Comment by sirbrillig** [comment]
Looks right to me based on what we discussed. I think we should also add `locale` to the `propTypes` for the survey step since we're relying on it being there.

I tried testing it, but it doesn't seem to solve the bug. The first switch (eg: en → fr) works, but then the second switch (eg: fr → en) does not translate the list items correctly. It *does* re-order them, though, which means that the array is being re-fetched. It's just not preventing the translations from failing.

I tried several things in the i18n module but I couldn't figure out how to get it to work. Was it working for you?

**Comment by kwight** [comment]
Ha, no, that was exactly what I was getting too (I should have mentioned the original description was the result of my not-working PR).

**Comment by kwight** [comment]
Looking at `blame`, maybe @rralian or @deBhal could chime in with some insights?

**Comment by deBhal** [comment]
It sounds like something's going wrong with i18nState in the i18n module.  I'll get in and debug this tomorrow, but in the meantime, there's some special casing around 'en' - is it any second change of locale or is 'en' special?

`reRenderTranslations()` doesn't change any of the translation data, it literally just rerenders all the translations.  We added this hook for translation plugins, e.g. the Community Translator triggers it when it turns on or off, so that i18n can run all the translations through the plugins.

You shouldn't need to do it here.  The i18n mixin should emit a change event and rerender everything when the locale is changed in `setLocale()`.

I might be missing some important context here.  Where is the props change coming from?

**Comment by rralian** [comment]
I find the code in signup to be difficult to follow. One of the benefits of React is that you can simplify a component down to only worrying about the props arguments you pass into it (and its internal state, which we try to minimize). But there are various steps here that make it somewhat difficult to see which components are being used and where the props are coming from.

@deBhal is correct in that `reRenderTranslations()` is unnecessary here. The key is to find how and when the `locale` prop changes in `stepsSurvey`. That component is required through `/client/signup/config/step-components.js`, which is required through `/client/signup/main` and assigned to `CurrentComponent`. `CurrentComponent` doesn't have `locale` passed directly, but rather through `propsFromConfig`, which comes from _either_ `this.props` or `steps[ this.props.stepName ].props`. I haven't run this branch, just looking at the code, so I believe we're probably getting this through `this.props` which would be coming from `client/signup/controller.js` when we pass `locale` into `SignupComponent` The `locale` prop is assigned once when `start()` is initially run. There is no means by which any updated value for `locale` would be passed into `SignupComponent`. I'm surprised the translations work the first time.

I think it would make things much simpler for this case to just run the translations on rendering of the component `/client/signup/steps/survey/index.jsx` and pass that around as an object rather than assigning them to state on mounting and then trying to change them when the locale changes. The `i18n` mixin handles that for you and will force a re-render when the locale changes. This means translations are re-run whenever the component is re-rendered, but I think it's better to keep this simple. If you find the component re-renders a lot, it's better to solve that problem directly.

**Comment by sirbrillig** [comment]
The reason for using the local state is that the order of the translations is randomized each time the user visits that signup step, but since this step includes two parts, if the user clicks on a tier-1 category, then changes their mind and uses the built-in back button to go back to the tier-1 list, we decided that it would be best if the order of categories didn't change. So we store the order in local state for the duration of signup.

If that state is causing the problem here, then we can try keeping the translations separate from the order of categories and hope that then `translate()` will have the correct locale set, but I'm skeptical that it will have any effect. That is, we could keep all the actual category data, including translations, in an object keyed by the category ID, then keep an array of category IDs which is what gets shuffled by `verticals.get()`.

File: client/signup/steps/survey/index.jsx [hunk 1]
```diff
@@ -14,6 +14,7 @@ import Card from 'components/card';
 import CompactCard from 'components/card/compact';
 import BackButton from 'components/header-cake';
 import Gridicon from 'components/gridicon';
+import i18n from 'lib/mixins/i18n';
```
Comment (Tug) [inline client/signup/steps/survey/index.jsx]: Hi there, i18n usage has changed recently. please use:
```
import i18n from 'i18n-calypso';
```
instead.

**Comment by lancewillett** [comment]
Reviewing older pull requests and branches, found this one — looks like it's abandoned. @kwight @deBhal do you want to continue with this fix? Related bug #4741 is still valid.

**Comment by lancewillett** [comment]
Ping @kwight to see if still in progress.

**Comment by kwight** [comment]
Both signup and i18n have gone through significant changes since looking at this; let's close for now, and we can reopen if someone decides they can take it on again.

**Pull request closed without merging**